corehq/apps/hqcase/bulk.py

```diff
@@ -1,9 +1,32 @@
 import time
+from dataclasses import dataclass
 from xml.etree import cElementTree as ElementTree
 
+from corehq.apps.users.util import SYSTEM_USER_ID, username_to_user_id
 from corehq.form_processor.interfaces.dbaccessors import CaseAccessors
 
-from .utils import CASEBLOCK_CHUNKSIZE, submit_case_blocks
+from .utils import CASEBLOCK_CHUNKSIZE, SYSTEM_FORM_XMLNS, submit_case_blocks
+
+
+@dataclass(frozen=True)
+class SystemFormMeta:
+    user_id: str = SYSTEM_USER_ID
+    username: str = SYSTEM_USER_ID
+    device_id: str = SYSTEM_USER_ID
+    xmlns: str = SYSTEM_FORM_XMLNS
+
+    @classmethod
+    def for_script(cls, name, username):
+        user_id = username_to_user_id(username)
+        if not user_id:
+            raise Exception(f"User '{username}' not found")
+
+        return cls(
+            user_id=user_id,
+            username=username,
+            device_id=name,
```

Comment:
Any particular reason that both the device id and the xmlns reference the script name? I don't feel strongly, just curious.

Reply:
No, and I'd welcome input with this sort of thing.  Convention in the `covid` migration scripts is to set `user_id`, but of course `username` is useful too.  Scripts of this sort do commonly set `device_id` to refer to the place this came from, but that's not easily queryable, at least not in reports.  I figured there's no _disadvantage_ to setting both `device_id` and `xmlns` to something unique to the script, so I just set each to the most obvious thing.

Reply:
Totally agreed about setting a convention. And agreed that there's no harm in setting both device and xmlns.

Regarding the user, it'd be nice to be able to default to the system user. I think that's actually a comment on the covid script - why is username required now?

Reply:
Oh, I thought we were always using our Dimagi users to indicate who applied a particular change, so I just made it mandatory.  If delivery and the partner don't care from an auditability perspective then I agree it'd be nicer to just use system.

Reply:
Yeah, that was my takeaway

Reply:
The idea about setting xmlns is interesting, I thought we could filter for system forms, but this makes it possible to filter for a specific form.
How does this look in the UI though or it's not visible yet because its not a part of the application?

Also, should we use the full path instead of just the last part of the name, so it does not conflict with another and is also easy to find? Looks like this script has already been running but since we are talking about a convention to follow.

Reply:
In the UI I believe this would come through as an unknown xmlns - I've struggled to get `FormsByApplicationFilter` to do anything I want it to besides show standard application forms, so I don't know how to force SubmitHistory to filter on this in the UI.  The main thing I had in mind here was that this should make it easier to follow up on the script in code, such as if there were an error we needed to undo.

I was thinking of this classmethod at least being used for management commands, where the full path is super long and the filename already exists in a global namespace (`./manage.py my_command` doesn't specify a django app).

Reply:
> so I don't know how to force SubmitHistory to filter on this in the UI

Cool. I think it's good to not have them appear in UI till we have a good way of displaying it to the user in a way that they understand what it is.

> I was thinking of this classmethod at least being used for management commands, where the full path is super long and the filename already exists in a global namespace

can you confirm what the xmlns and device id appear in the forms you are submitting now through the command in this PR?

Reply:
device id: `custom.covid.management.commands.update_all_activity_complete_date`
xmlns: `http://commcarehq.org/script/update_all_activity_complete_date`

Reply:
Thank you @esoergel

Reply:
:+1: oddly enough though - it looks like viewing the form metadata in the UI doesn't display the correct XMLNS - it shows the `xmlns:orx` from the template, not the `xmlns`:

https://github.com/dimagi/commcare-hq/blob/567e48f97c0cb0130f4a03e4a64bdf73c981f9f3/corehq/apps/hqcase/templates/hqcase/xml/case_block.xml#L2-L2

However, when viewed in `raw_data` you can see that `form.xmlns` and `form.form['@xmlns']` both reference the proper one. :shrug: